LUN -1033

laurafeier opened the pull request:
Unfiled files folder:
- files cannot be added to this folder
- files can only be moved out of this folder 

Core Folders: 
- files cannot be changed; all fields are readonly
- there should be no way of moving/copy files inside them
- no actions are allowed on core files/folders
- users can only view and use them
- there is no way of creating a core folder through the admin

Site Folders:
- are required to have a site assigned
- can only be changed by superusers or site admins
- site ownership can only be changed from root folders
- on site change sub-folders will have the same site assigned
- only superusers and site admins can change root site folder
- when there is no site ownership, only superusers can delete/move them 

Restricted Folders:
- only site folders can be restricted/unrestricted
- only superusers and users that have role with groups that have permission 'can_restrict_operation' assigned(site admins/developers) can restrict/unrestrict
- users like editors and writers should not be able to change anything from the asset that is readonly
- users like editors and writers should not be able to delete/move a folder if that folder contains restricted assets 

Shared Folder:
- only superusers can share folder with other sites
- files cannot be changed; all fields are readonly
- there should be no way of moving/copy files inside them
- no actions are allowed 
- users can only view and use them in pages

Fixes:
- cannot move/copy files to a destination that already contains them (duplicate validation)
- files that failed to be moved from clipboard are no longer deleted from clipboard
- since there is no way to find the current folder when searching, actions will be disabled

Commit:
LUN-1033: added folder type field.

Commit:
Added site field for folder.

Commit:
* merged all add folder forms into one.
* moved folder duplication form validation in model validation
* added restrictions for site folders
* some pep8 & pylint
* removed the make_folder view and added new one that uses
	django's add view
* site for site folders can be changed only on root folders
* the site for root site folders are set(on save) for all
	subfolders

Commit:
Added folder type display name in directory-listing template.

Commit:
Merge remote-tracking branch 'origin/master_pbs' into LUN-1033

Conflicts:
	test_settings.py

Commit:
Merge remote-tracking branch 'origin/master_pbs' into LUN-1033

Commit:
New tests - unfinished. New folder type permission rules.

Commit:
Added some more permissions tests

Commit:
Added tests for moving files based on filer type permissions layer
Changed file/folder manager to support chainable querysets

Commit:
Move/copy actions now excludes core folders from destinations
Refactored admin tests.

Commit:
* Core folder files => readonly + no submission in view;
* Restricted permissions in core folders for 'single file'
	move to clipboard;
* Added file name validation on move action;
* Added restricted on extracting archive files;
* Files with name confliction on 'move from clipboard' will not get
	deleted anymore;
* Some folder functions/buttons will no loger appear in template
	if folder is core folder ;
* Code refactoring in admin tests;
* Added some helper functions for admin tests;

Commit:
stupid 'self' instead of awesome 'TestCase'

Commit:
*extracting archives no longer available in: core or unfiled files;
*archive zip validation detects malformed files;
* added tests for folder save functionality

Commit:
Removed FILER_ALLOW_REGULAR_USERS_TO_ADD_ROOT_FOLDERS since there
	is no more functionality for it.

zoliszeredi commented on filer/models/filemodels.py:
```diff
@@ -27,10 +31,30 @@ def find_all_duplicates(self):
         return r
 
     def find_duplicates(self, file_obj):
```
Since the class name is renamed to FilesChainableQuerySet, shouldn't it return query set? find_duplicates returns a list and find_all_duplicates returns a dict.

zoliszeredi commented on filer/models/filemodels.py:
```diff
@@ -10,13 +10,17 @@
+        return self.filter(folder__folder_type=Folder.CORE_FOLDER)
 
-class FileManager(PolymorphicManager):
     def find_all_duplicates(self):
```
Same as below.

zoliszeredi commented on filer/tests/admin.py:
```diff
@@ -542,3 +612,683 @@ def test_admin_queries(self):
+        self.assertEqual(response.status_code, 403)
+        response = self.client.post(
+            reverse('admin:filer_folder_change', args=(f1.pk, )), {})
+        self.assertEqual(response.status_code, 403)
```
You are testing the change on both core folders and site folders in the same test. If the first assert fails the other case won't have test coverage. How about splitting up this test into two?

zoliszeredi commented on filer/tests/admin.py:
```diff
@@ -542,3 +612,683 @@ def test_admin_queries(self):
+            reverse('admin:filer_folder_change', args=(f1.pk, )), {})
+        self.assertEqual(response.status_code, 200)
+
+    def test_delete_root_folder(self):
```
Same comment as for the change root folder test.

Commit:
* impoved performance on find_all_duplicates method.
* find_duplicates method returns now queryset instead of list

Commit:
Removed folder permission model and all its logic.

Commit:
Added django-cms-roles as a dependency for django-filer.
THe permission management will be build on top of cmsroles.

Commit:
SitFolder: viewable by users that have access on the folder's site

Commit:
* moved views methods in the specific model admins;
* added new admin tools to detect if user has permissions
	for specific actions
* added site filtering based on user roles over sites
* action checkboxes will not be available for core files
	and folders;
* actions will not be available when doing searches
* replaced all permission logic
* left all disabled actions commented and mostly unchanged;
	these will need to be changed if they will be
	enabled again

Commit:
Removed unused tests for old permissions.

Commit:
Changed restricted name methods to readonly since restricted will
	be a new feature.

Commit:
Merge remote-tracking branch 'origin/master_pbs' into LUN-1033

Commit:
LUN-1041: removed focal point feature.

Commit:
LUN-1116: Added functionality for the rule that users cand see
only current site's folders when editing pages.

Commit:
LUN-1116: current site folder filtering when selecting filer objs.
* actions for files/folders are now hidden when selecting filer
	objects for plugins
* actions available: make subfolders, upload files, move from clip
* changed redirect for file/folder admin to be the same

Commit:
LUN-1118: generated deletion link was incorrect for file subtypes

Commit:
Forbidden is thrown if trying to change a file without permission
	on its folder.

Commit:
Fixed change image to redirect to parent folder. Url for redirect
	was generated by django based on the general change perm.
* created method for duplicated code that redirects to dir-listing

Commit:
Actions move/delete are now hidden for users that do not have
	'can_delete' permission assigned.

Commit:
Actions move/delete are available for unfiled files.
Action extract should not be available in unfiled files folder
	since files from that folder should only move out and
	never get more of them. All files need to have a site
	ownership.

Commit:
LUN-1119: fixed search limited to current folder to include files
	from the current folder.

Commit:
LUN-1121: improved performance by minimising number of queries on
	directory listing view.

Commit:
* disable selection for all files and folders since this will
	include the unfiled files folder and it shouldn't

Commit:
* refactored template code for directory table;
* moved listing for virtual folders in template code

Commit:
LUN-1122: fixed search count
* minimised no of queries for folder children files/folders count

Commit:
* added MultiMoldelQuerysetChain utility to make sure paginator
does not load all objects from querysets in memory.
* removed select related since it generates too many LEFT OUTER
JOIN statements and even the number of queries are less, it takes
a larger amount of times to load.

Commit:
* fixed tests to exclude unfiled folder since it will no loger
	appear

Commit:
* code refactoring for more easier further development

Commit:
Changed the way file permissions are handled:
* writers/editors will be able to delete or change metadata of the
	files they own only if asset belongs to the site they have
	role/s on
* writers/editors will be able to see other users files from the
	site they have role on but as readonly

Commit:
Make sure uploaded files are closed even if error is raised

Commit:
Fixed string conversion for error message on duplicates check.

Commit:
* reverted permission changes for editor/writer roles
* fixed logic for Back button when changing files in popup
	dir-list view
* added helper tag to build popup parameters for requests

Commit:
* raise 404 instead of 500 when obj does not exists

Commit:
* fixed change folder url

Commit:
LUN-1007: added restricted fields for files and folders

Commit:
* restricted field is readonly for users that do not have perms to
	change it

Commit:
* users can now make restricted assets
* if folder is restricted=> all descendants will be restricted
* restricted assest can get un-restricted only from the parent
that was set to restricted

Commit:
* readonly admin field styled when readonly

Commit:
* preserve restriction on save from parent if parent is restricted

Commit:
Added permissions checks for restricted fields.

Commit:
* user without permission restrictions will not be able to see
	change links for restricted files/folders

Commit:
* removed restircted folders from destination folder list

Commit:
* dir list actions are now hidden for restricted folders
* move files from clipboard to restricted folder is now disabled

Commit:
* limit_search_to_folder will not be visible in root folders

Commit:
* writer user on site x can only do operations on site x folders
	even if it is site-admin on site y
* can restrict permission check is now done based on
	site from folder

Commit:
* allow make folder in root for site admins

Commit:
* allow superusers to do actions on restricted folders

Commit:
* fixed restricted queryset filtering to exclude all files/folders
	if there are no sites without restriction permission

Commit:
* hide change folder link for user that don't have permission 4 it

Commit:
*new actions enable/disable restriction are available in filer

Commit:
* changed tests to assert for the success of an operation not for
	response code

Commit:
* changed owner field display to show email(if there is one)

Commit:
* filtered unfiled files that belong to someone's clipboard

Commit:
* added new field for folder: shared
* shared formfield will only be available for superuser
* regenerated migrations to include a datamigration for binding
	sites to folders

Commit:
* renamed readonly functions since the concept for readonly assets
	will be applied for core folders and shared folders

Commit:
* make folder button was missing in root folder since method
	is_readonly_for_user was used in template but was not
	defined for root virtual folder

Commit:
* restriction - keep from parent only if parent is restricted

Commit:
* fixed enable/disable restriction actions showing up for children
	selection when only current folder can be unrestricted
* fixed numeric conversion for site_id to work for floats too

Commit:
* fixed unfiled files from clipboard exclusion

Commit:
extract files should not be available for any virtual folders

Commit:
* fixed tests for actions on unfiled files;
* restricted destinations are now disabled

Commit:
* added restriction permission check in all descendants
* excluded restricted destination subfolders from available
	destination list
* added tests for admin/tools.py coverage

Commit:
* added restriction validation to be available in delete_view
* site admins should not be able to delete folders with no site

Commit:
* added perms check for destination when cloning clipboard

Commit:
* added helper decorator for cms_roles functions to get rid of
	the duplicated code

Commit:
LUN-1154: added the following functionality for shared folders:
* users that have role on site X can now see other sites folders
if those are shared with site X
* users from site X can use shared folders assets on their pages
* users from site X can only view shared folders assets

Commit:
* added test for shared folders functionality

Commit:
* fix: shared sites from root will get propagated to descendants
* added more tests to make sure all descendants get updated and
that only root folders can be shared from the admin

Commit:
LUN-1164: fixed recursion error when copy folder to itself

Commit:
* changed the way folders are saved on move/copy since mptt
methods move_to, insert_at can generate tree coruptions. Changing
only the parent folder seems safer since folders/files are always
shown to the user ordered by name and their position in tree
doesn't matter
* changed fks for owner and site so that on deletion folders/files
won't get deleted from filer
* fixed descendants updating metadata only when metadata has
changed

Commit:
* fixed new folders to inherit shared sites from parent

Commit:
* in order to not rely on the descendants query from mptt(due to
tree corruptions) added a new method that gets all descendants
recusive by getting all children of children of children ...

Commit:
* fix: shared sites were not propagated to all descendants on
	parent change

Commit:
* folder change form template should use the same breadcrumbs
template as the rest of the filer templates

Commit:
LUN-1173: fixed most of the operations that generated
tree corruptions

Commit:
* removed unused method from tests(got there by mistake)

Commit:
* new command to fix/check tree corruptions available

Commit:
* removed duplicated code since current_site var will be available
 to all js files included in the template that has current_site
context variable

Commit:
LUN-1177: implemented data migration to map a site for each
existing folder based on its folder permissions and metadata
(the guessing functions are based on the string similarities
 between site names and folder name/owner email)

bhodorog commented on filer/migrations/0017_assign_sites_to_folders.py:
```diff
@@ -0,0 +1,498 @@
+        self.init_logger()
+        self.fix_tree()
+
+        self.folders_migrated = set()
```
using this as global attribute, modifiable by self.make_core_folders, self.map_defined_folders is not readable. Is hard to guess who's adding to that set.

bhodorog commented on filer/migrations/0017_assign_sites_to_folders.py:
```diff
@@ -0,0 +1,498 @@
+from cmsroles.models import Role
+
+
+class Migration(DataMigration):
```
a doc comment explaining the idea behind this migration would be helpful.
